### PR TITLE
[`PPOTrainer`] Support generic optimizers

### PR DESCRIPTION
**Pull request by younesbelkada**
This PR adds the support of generic optimizers. Before this PR, the `PPOTrainer` was only support adam optimizer. Users are now free to use any optimizer.
Added also an example that leverages `8bitAdam` which is lighter & faster than classic `Adam` optimizer. 

cc @lewtun @lvwerra @edbeeching 

as a side note, 8bitAdam should support DP out of the box

### CODE REVIEW

**Comment by younesbelkada** [comment]
Regarding 8-bit Adam, it is quite hard to make it converge. I have found that the model falls rapidly in a collapse mode: https://wandb.ai/distill-bloom/trl/runs/k7vogzao?workspace=user-younesbelkada let me know if it still makes sense to add the example

File: trl/trainer/ppo_trainer.py [hunk 1]
```diff
@@ -70,6 +71,8 @@ def __init__(
             dataset (Union[`torch.utils.data.Dataset`, `datasets.Dataset`]):
                 PyTorch dataset or Hugging Face dataset. If a Hugging Face dataset is passed, the dataset
                 will be preprocessed by removing the columns that are not used by the model.
+            optimizer (Optional[`torch.optim.Optimizer`]):
+                Optimizer used for training. If `None`, the default optimizer is used.
```
Comment (lvwerra) [inline trl/trainer/ppo_trainer.py:75]:
```suggestion
                Optimizer used for training. If `None`, the `Adam` is used as default.
```

**Review by lvwerra — COMMENTED** [review]
Looks good to me, just a small nit. Do you want add the `scheduler` here or in a new PR?

**Comment by younesbelkada** [comment]
Let's address the scheduler in a follow up PR!

**Review by lvwerra — APPROVED** [review]

**Comment by LouisCastricato** [comment]
@younesbelkada FYI, 8bit adam converges only after you do a lot of stuff with reward normalization. https://github.com/CarperAI/trlx/issues/53 see here. We also had significant issues getting it working. There was also a recent bug in computing advantages that we found that I believe was carried over from TRL, I'll have to double check with one of my engineers on this.

**Comment by LouisCastricato** [comment]
Nevermind, it appears like the bug is a non issue for TRL.
